presto-main-base/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java

```diff
@@ -1685,6 +1686,54 @@ public static double arrayL2SquaredDouble(
         return sum;
     }
 
+    @Description("Dot Product distance between the given identical sized vectors represented as DOUBLE arrays")
+    @ScalarFunction("dot_product")
+    @SqlNullable
+    @SqlType(StandardTypes.DOUBLE)
+    public static Double arrayDotProduct(
+            @SqlType("array(double)") Block leftArray,
+            @SqlType("array(double)") Block rightArray)
+    {
+        checkCondition(
+                leftArray.getPositionCount() == rightArray.getPositionCount(),
+                INVALID_FUNCTION_ARGUMENT,
+                "Both array arguments must have identical sizes");
+        double result = 0.0;
+
+        for (int i = 0; i < leftArray.getPositionCount(); i++) {
+            if (leftArray.isNull(i) || rightArray.isNull(i)) {
+                return null;
+            }
+            result += DOUBLE.getDouble(leftArray, i) * DOUBLE.getDouble(rightArray, i);
+        }
```

Comment:
I think you'll have better performance by removing the data dependency in the loop.

```suggestion
        for (int i = 0; i < leftArray.getPositionCount(); i++) {
            result += DOUBLE.getDouble(leftArray, i) * DOUBLE.getDouble(rightArray, i);
        }
        
        boolean isNull = false;
        if (leftArray.mayHaveNulls() || rightArray.mayHaveNulls()) {
            for (int i = 0; i < leftArray.getPositionCount(); i++) {
                isNull = isNull || leftArray.isNull(i) || rightArray.isNull(i);
            }
        }
        if (isNull) {
            return null;
        }
        return result;
```